_posts/uportal/2017-12-06-myuw-fork.md

```diff
@@ -0,0 +1,168 @@
+`uPortal-home` the requests that can be handled by `uPortal home`. The 
+traditional rendering pipeline still renders portlets in maximized and 
+exclusive window states.
+
++ Prefer to render static content via `/web/static/{fname}`
++ Prefer to render portlets via `/web/exclusive/{fname}` (`renderOnWeb`)
++ Bounce attempts to do anything other than render just one portlet (i.e. maximized or exclusive) to `/web`
++ Coerce guest user to the desired profile
+
+# Hacked theme
+
+MyUW uses a customized "Bucky" theme forked from Respondr. The purpose of this customized theme and associated skins is to make rendering a portlet maximized via uPortal look like rendering uPortal-home, so that users switch reasonably seamlessly between the two approaches depending on content.
+
++ Fewer regions. Simpler. Remove features MyUW isn't using.
++ When rendering a maximized portlet, simulate the way uPortal-home looks, so that rendering maximized portlets feels seamless enough. That is, stylize the header.
++ On add-to-home, clear uPortal-home-specific session storage items so that change will reflect when rendering `uPortal-home`
+
+
+# Hacked JSON APIs
+
+MyUW has modified JSON APIs to expose data `uPortal-home` needs.
+
++ Expose more data about portlet publications, to power widgets. e.g. `renderOnWeb`,`mdIcon`, `widgetURL`, `widgetType`, `widgetConfig`, `widgetTemplate`, `widgetExternalMessageUrl`, `widgetExternalMesageTextObjectLocation`, `widgetExternalMessageLearnMoreUrlLocation`, in `LayoutPortlet`. See also idea to use portlet publishing parameters for this instead.
++ Add `/marketplace/entry/{fname}.json` to get JSON for a single Marketplace entry
++ Add `/portlet/{fname}.json` to get JSON for a single portlet
++ Add `GroupRESTController`
++ Defaulting `url` of LayoutPortlet to that to render the portlet via uPortal
++ Add `lifecycleState` to `LayoutPortlet`
++ Add simplified move-portlet API call (`action=movePortletAjax` in `UpdatePreferencesServlet`)
++ Add remove-from-layout-by-fname (`action=removeByFName` in `UpdatePreferencesServlet`)
++ Add `firstName`, `lastName` in the `SessionRESTController` response. Also, `originalUsername`  when swapped.
++ Add some error handling in `LayoutPortlet` to catch when `widgetConfig` is not valid JSON.
++ Increase time-to-live of `marketplacePortletDefinitionCache`
++ Filter marketplace entries to only those in `PUBLISHED` publication lifecycle state. (Arguably this should be more nuanced, reflecting the viewing user's permissions -- maybe you've got permission to see entries that are in states other than `PUBLISHED`).
++ Attempts to make hitting JSON APIs have the side effect of logging the user in / bootstrapping a uPortal session when they can. This is fraught.
+
+# Hacked DLM
+
+MyUW has modified DLM to simplify the user experience of adding content to one's homepage.
+
++ Default adding a new item to the first rather than last position in a container.
```

Comment:
There was an effort to upstream this in https://github.com/Jasig/uPortal/pull/885 the PR stalled before it could be merged. :disappointed: